### PR TITLE
Mucking around with scalajs

### PR DESCRIPTION
**Pull request by snoble**
JS builds of core works. The new cli target does not

### CODE REVIEW

**Comment by johnynek** [comment]
I don't think EnumValue or Fn need to be defined in Java. I did that originally to make java interop easier, but it wasn't thought out very much. I think if it makes making a js/native version any easier, we should just make those scala.

**Comment by snoble** [comment]
Interesting. In that case the Java path stuff might be abstracted out into a small class of its own

**Comment by snoble** [comment]
still have to sort out how to deal with `Path` which leaks out of `Main` into a few other places

**Comment by snoble** [comment]
also the json tests use jawn which doesn't seem to exist for scalajs

**Comment by johnynek** [comment]
we can probably sort the that out by having a separate module that is standard scala only.

**Comment by snoble** [comment]
Did a little bit of stuff with making `Path` work cross platform. Still need some file reading/writing for js (for when it's used with node). But right now the errors get stuck on reflection FFI code

**Comment by snoble** [comment]
This might be getting to a place that's too hairy for me to get through. Namely, a lot of the FFI code uses stuff that's not supported in a trivially way in scalajs

An example of some errors
```
[error] Referring to non-existent class java.lang.Class$
[error]   called from org.bykn.bosatsu.FfiCall$Reflective.$$anonfun$call$3(org.bykn.bosatsu.rankn.Type,scala.collection.immutable.List,java.lang.String,scala.Function1)org.bykn.bosatsu.Evaluation$Value
[error]   called from org.bykn.bosatsu.FfiCall$Reflective.call(org.bykn.bosatsu.rankn.Type)cats.Eval
[error]   called from org.bykn.bosatsu.Evaluation.$$anonfun$eval$1(scala.Tuple2,scala.Function1)scala.Tuple2
[error]   called from org.bykn.bosatsu.Evaluation.<init>(org.bykn.bosatsu.PackageMap,org.bykn.bosatsu.Externals)
[error]   called from org.bykn.bosatsu.TestUtils$.inferredHandler$1(org.bykn.bosatsu.PackageMap,org.bykn.bosatsu.PackageName,org.bykn.bosatsu.TestUtils$EvaluationMode,org.bykn.bosatsu.Externals)org.scalatest.compatible.Assertion
[error]   called from org.bykn.bosatsu.TestUtils$.$$anonfun$evalTestMode$5(org.bykn.bosatsu.TestUtils$EvaluationMode,org.bykn.bosatsu.Externals,org.bykn.bosatsu.PackageMap,org.bykn.bosatsu.PackageName)org.scalatest.compatible.Assertion
[error]   called from org.bykn.bosatsu.TestUtils$.evalTestMode(scala.collection.immutable.List,java.lang.String,org.bykn.bosatsu.TestUtils$EvaluationMode,org.bykn.bosatsu.Externals)org.scalatest.compatible.Assertion
[error]   called from org.bykn.bosatsu.TestUtils$.runBosatsuTest(scala.collection.immutable.List,java.lang.String,scala.Int,org.bykn.bosatsu.Externals)org.scalatest.compatible.Assertion
[error]   called from org.bykn.bosatsu.EvaluationTest.<init>()
[error]   called from org.bykn.bosatsu.EvaluationTest.__exportedInits
[error]   exported to JavaScript with @JSExport
[error] involving instantiated classes:
[error]   org.bykn.bosatsu.FfiCall$ScalaCall
[error]   org.bykn.bosatsu.FfiCall$JavaCall
[error]   org.bykn.bosatsu.Evaluation
[error]   org.bykn.bosatsu.TestUtils$
[error]   org.bykn.bosatsu.EvaluationTest
[error] Referring to non-existent method java.lang.Class$.forName(java.lang.String)java.lang.Class
[error]   called from org.bykn.bosatsu.FfiCall$Reflective.$$anonfun$call$3(org.bykn.bosatsu.rankn.Type,scala.collection.immutable.List,java.lang.String,scala.Function1)org.bykn.bosatsu.Evaluation$Value
[error]   called from org.bykn.bosatsu.FfiCall$Reflective.call(org.bykn.bosatsu.rankn.Type)cats.Eval
[error]   called from org.bykn.bosatsu.Evaluation.$$anonfun$eval$1(scala.Tuple2,scala.Function1)scala.Tuple2
[error]   called from org.bykn.bosatsu.Evaluation.<init>(org.bykn.bosatsu.PackageMap,org.bykn.bosatsu.Externals)
[error]   called from org.bykn.bosatsu.TestUtils$.inferredHandler$1(org.bykn.bosatsu.PackageMap,org.bykn.bosatsu.PackageName,org.bykn.bosatsu.TestUtils$EvaluationMode,org.bykn.bosatsu.Externals)org.scalatest.compatible.Assertion
[error]   called from org.bykn.bosatsu.TestUtils$.$$anonfun$evalTestMode$5(org.bykn.bosatsu.TestUtils$EvaluationMode,org.bykn.bosatsu.Externals,org.bykn.bosatsu.PackageMap,org.bykn.bosatsu.PackageName)org.scalatest.compatible.Assertion
[error]   called from org.bykn.bosatsu.TestUtils$.evalTestMode(scala.collection.immutable.List,java.lang.String,org.bykn.bosatsu.TestUtils$EvaluationMode,org.bykn.bosatsu.Externals)org.scalatest.compatible.Assertion
[error]   called from org.bykn.bosatsu.TestUtils$.runBosatsuTest(scala.collection.immutable.List,java.lang.String,scala.Int,org.bykn.bosatsu.Externals)org.scalatest.compatible.Assertion
[error]   called from org.bykn.bosatsu.EvaluationTest.<init>()
[error]   called from org.bykn.bosatsu.EvaluationTest.__exportedInits
[error]   exported to JavaScript with @JSExport
[error] involving instantiated classes:
[error]   org.bykn.bosatsu.FfiCall$ScalaCall
[error]   org.bykn.bosatsu.FfiCall$JavaCall
[error]   org.bykn.bosatsu.Evaluation
[error]   org.bykn.bosatsu.TestUtils$
[error]   org.bykn.bosatsu.EvaluationTest
[error] Referring to non-existent method java.lang.Class.getMethod(java.lang.String,[java.lang.Class)java.lang.reflect.Method
[error]   called from org.bykn.bosatsu.FfiCall$Reflective.$$anonfun$call$3(org.bykn.bosatsu.rankn.Type,scala.collection.immutable.List,java.lang.String,scala.Function1)org.bykn.bosatsu.Evaluation$Value
[error]   called from org.bykn.bosatsu.FfiCall$Reflective.call(org.bykn.bosatsu.rankn.Type)cats.Eval
[error]   called from org.bykn.bosatsu.Evaluation.$$anonfun$eval$1(scala.Tuple2,scala.Function1)scala.Tuple2
[error]   called from org.bykn.bosatsu.Evaluation.<init>(org.bykn.bosatsu.PackageMap,org.bykn.bosatsu.Externals)
[error]   called from org.bykn.bosatsu.TestUtils$.inferredHandler$1(org.bykn.bosatsu.PackageMap,org.bykn.bosatsu.PackageName,org.bykn.bosatsu.TestUtils$EvaluationMode,org.bykn.bosatsu.Externals)org.scalatest.compatible.Assertion
[error]   called from org.bykn.bosatsu.TestUtils$.$$anonfun$evalTestMode$5(org.bykn.bosatsu.TestUtils$EvaluationMode,org.bykn.bosatsu.Externals,org.bykn.bosatsu.PackageMap,org.bykn.bosatsu.PackageName)org.scalatest.compatible.Assertion
[error]   called from org.bykn.bosatsu.TestUtils$.evalTestMode(scala.collection.immutable.List,java.lang.String,org.bykn.bosatsu.TestUtils$EvaluationMode,org.bykn.bosatsu.Externals)org.scalatest.compatible.Assertion
[error]   called from org.bykn.bosatsu.TestUtils$.runBosatsuTest(scala.collection.immutable.List,java.lang.String,scala.Int,org.bykn.bosatsu.Externals)org.scalatest.compatible.Assertion
[error]   called from org.bykn.bosatsu.EvaluationTest.<init>()
[error]   called from org.bykn.bosatsu.EvaluationTest.__exportedInits
[error]   exported to JavaScript with @JSExport
[error] involving instantiated classes:
[error]   org.bykn.bosatsu.FfiCall$ScalaCall
[error]   org.bykn.bosatsu.FfiCall$JavaCall
[error]   org.bykn.bosatsu.Evaluation
[error]   org.bykn.bosatsu.TestUtils$
[error]   org.bykn.bosatsu.EvaluationTest
[error] Referring to non-existent method java.lang.Class.getDeclaredField(java.lang.String)java.lang.reflect.Field
[error]   called from org.bykn.bosatsu.FfiCall$Reflective.$$anonfun$call$1(java.lang.Class)java.lang.Object
[error]   called from org.bykn.bosatsu.FfiCall$Reflective.call(org.bykn.bosatsu.rankn.Type)cats.Eval
[error]   called from org.bykn.bosatsu.Evaluation.$$anonfun$eval$1(scala.Tuple2,scala.Function1)scala.Tuple2
[error]   called from org.bykn.bosatsu.Evaluation.<init>(org.bykn.bosatsu.PackageMap,org.bykn.bosatsu.Externals)
[error]   called from org.bykn.bosatsu.TestUtils$.inferredHandler$1(org.bykn.bosatsu.PackageMap,org.bykn.bosatsu.PackageName,org.bykn.bosatsu.TestUtils$EvaluationMode,org.bykn.bosatsu.Externals)org.scalatest.compatible.Assertion
[error]   called from org.bykn.bosatsu.TestUtils$.$$anonfun$evalTestMode$5(org.bykn.bosatsu.TestUtils$EvaluationMode,org.bykn.bosatsu.Externals,org.bykn.bosatsu.PackageMap,org.bykn.bosatsu.PackageName)org.scalatest.compatible.Assertion
[error]   called from org.bykn.bosatsu.TestUtils$.evalTestMode(scala.collection.immutable.List,java.lang.String,org.bykn.bosatsu.TestUtils$EvaluationMode,org.bykn.bosatsu.Externals)org.scalatest.compatible.Assertion
[error]   called from org.bykn.bosatsu.TestUtils$.runBosatsuTest(scala.collection.immutable.List,java.lang.String,scala.Int,org.bykn.bosatsu.Externals)org.scalatest.compatible.Assertion
[error]   called from org.bykn.bosatsu.EvaluationTest.<init>()
[error]   called from org.bykn.bosatsu.EvaluationTest.__exportedInits
[error]   exported to JavaScript with @JSExport
[error] involving instantiated classes:
[error]   org.bykn.bosatsu.FfiCall$ScalaCall
[error]   org.bykn.bosatsu.FfiCall$JavaCall
[error]   org.bykn.bosatsu.Evaluation
[error]   org.bykn.bosatsu.TestUtils$
[error]   org.bykn.bosatsu.EvaluationTest
[trace] Stack trace suppressed: run last coreJS/test:fastOptJS for the full output.
[error] (coreJS/test:fastOptJS) org.scalajs.core.tools.linker.LinkingException: There were linking errors
```

**Comment by johnynek** [comment]
Okay. We can consider just removing entirely reflective calls. I don’t think we use them at all. And reflection is objectively bad news (and also a problem for graal). Maybe just remove those classes and tests?

**Comment by snoble** [comment]
The tests are passing, but so slowly I figure I must have something setup wrong. eg
```
[info] ParserTest:                                                      
[info] - we can parse integers (405 milliseconds)                       
[info] - string escape/unescape round trips (645 milliseconds)          
[info] - we can parse quoted strings (390 milliseconds)                 
[info] - Identifier round trips (86 milliseconds)                       
[info] - we can parse lists (1 second, 77 milliseconds)                 
[info] - we can parse dicts (18 seconds, 846 milliseconds)              
[info] - we can parse tuples (760 milliseconds)                         
[info] - we can parse blocks (4 milliseconds)                           
[info] - we can parse TypeRefs (391 milliseconds)                       
[info] - we can parse python style list expressions (7 milliseconds)    
[info] - we can parse operators (599 milliseconds)                      
```
vs
```
[info] ParserTest:                                                        
[info] - we can parse integers (888 milliseconds)                         
[info] - string escape/unescape round trips (312 milliseconds)            
[info] - we can parse quoted strings (177 milliseconds)                   
[info] - Identifier round trips (103 milliseconds)                        
[info] - we can parse lists (362 milliseconds)                            
[info] - we can parse dicts (3 seconds, 18 milliseconds)                  
[info] - we can parse tuples (97 milliseconds)                            
[info] - we can parse blocks (18 milliseconds)                            
[info] - we can parse TypeRefs (62 milliseconds)                          
[info] - we can parse python style list expressions (15 milliseconds)     
[info] - we can parse operators (86 milliseconds)                         
```

**Comment by snoble** [comment]
I also introduced a macro into Predef... which is probably not done in the best way

**Comment by johnynek** [comment]
Scalajs is slow, so I’d expect the tests to take longer. Some of them too long maybe.

**Comment by snoble** [comment]
So it'll probably never be a great experience but it's nice to have as a way to release a static site that lets people play with the language without any server resources

**Comment by snoble** [comment]
I'll move this over to "Ready for review" just so CI can run and you can give a thought to if this is something you're interested in given the trade offs. I definitely hacked this in so don't hesitate to say big choices were wrong

**Comment by johnynek** [comment]
why do we need to move so many things into `shared`? Isn't that the default? E.g. https://github.com/typelevel/paiges/tree/master/core/src doesn't do that.

I would like to have scalajs merged since I also think it will keep us more honest in the compiler about not using crazy java stuff, which will also help native-image.

**Comment by snoble** [comment]
I think that's because they're using CrossType Pure, which is because they don't need any platform specific code.

https://github.com/typelevel/paiges/blob/5520666e87c93f4f275e4de90c53daf2cc133d5f/build.sbt#L53

Where as I did add some platform specific code around Path. This is similar to decline

https://github.com/bkirwi/decline/tree/master/core/shared/src

That said, the way I did platform specific isn't particularly well thought out. Just a way to push through to something working

**Comment by johnynek** [comment]
okay, yeah, I guess we do need to move to shared.... you are right.

And, I think it is fine to separate the CLI into a separate target that may not be js compatible. I think the JS code will be required to pass in a bunch of strings for now, which is what the browser would probably have access to anyway I guess...

We will need to update the bazel build since it depends on these paths.

**Comment by snoble** [comment]
That makes sense. Any thoughts on the externs param for the cli? That previously was where reflections was used. So they could be brought back if the cli is jvm only (though I don't know how that complicates Graal)

**Comment by johnynek** [comment]
I’m leaning towards totally removing reflective externs which means you’d have to compile a custom CLI if you want custom externs. This simplifies the scalajs and graal native-image (not to mention scala-native if for some reason you want that over native-image).

**Comment by johnynek** [comment]
What do you think of that?

**Comment by snoble** [comment]
Does it make any sense to make a library dependency that contains a Map of objects that can be used as external imports that gets imported by Evaluation?

**Comment by snoble** [comment]
@johnynek: I think if I pull out externals and rebase this will start to look pretty good

**Comment by snoble** [comment]
oh... and I still don't love the macro

**Comment by johnynek** [comment]
I actually do love the macro, I think it is a very nice solution to loading constants into source code. It may be an issue for very large constants, but I think we should be fine.

The bazel build will be a minor problem: we will have to declare that as data I think so bazel will let the compiler see the file, but should be doable.

**Comment by snoble** [comment]
Not sure how to best go out fixing the bazel build since I don't have that set up, but everything else seems to be working

**Comment by johnynek** [comment]
I can take that, but not till next week, I'm traveling this weekend.

but `brew install bazel` should work if you are on OSX. then `bazel build test_workspace/...` should work.

**Comment by snoble** [comment]
I'm getting closer, except I'm struggling with how to produce the equivalent of
```
lazy val scalaReflect = Def.setting { "org.scala-lang" % "scala-reflect" % scalaVersion.value }
```
in a bazel build

**Comment by snoble** [comment]
oh wait... `scala_macro_library` might be what I need

**Comment by snoble** [comment]
for some reason `bazel build //test_workspace/...` breaks on 
```
java.io.FileNotFoundException: core/src/main/resources/bosatsu/predef.bosatsu (No such file or directory)
        at java.io.FileInputStream.open0(Native Method)
        at java.io.FileInputStream.open(FileInputStream.java:195)
        at java.io.FileInputStream.<init>(FileInputStream.java:138)
        at scala.io.Source$.fromFile(Source.scala:91)
        at scala.io.Source$.fromFile(Source.scala:76)
        at scala.io.Source$.fromFile(Source.scala:54)
        at scala.io.Source$.fromFile(Source.scala:60)
        at org.bykn.bosatsu.Macro.smac(Macro.scala:10)

  private val predefString: String = f("core/src/main/resources/bosatsu/predef.bosatsu")
```
even though I can build `//core/...` and `//cli/...` just fine

File: base/src/main/scala/org/bykn/bosatsu/Macro.scala [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import scala.util.{Try => STry}
+
+class Macro(val c: Context) {
+  import c._, universe._
+  def smac(file: c.Expr[String]) = file.tree match {
+    case Literal(Constant(s: String)) =>
+      val res = STry(
+        Source.fromFile(s, "UTF-8").getLines.mkString("\n")
+      ).getOrElse(
+        Source.fromFile(s"external/org_bykn_bosatsu/$s", "UTF-8").getLines.mkString("\n")
```
Comment (snoble) [inline base/src/main/scala/org/bykn/bosatsu/Macro.scala:15]: I added this in just to get things working. We only end up over here when using `rules_bosatsu`. But I can't figure out why.

Comment (johnynek) [reply base/src/main/scala/org/bykn/bosatsu/Macro.scala:15]: The path to the file is different evidently in bazel and sbt. That’s a bummer.

Comment (johnynek) [reply base/src/main/scala/org/bykn/bosatsu/Macro.scala]: The path to the file is different evidently in bazel and sbt. That’s a bummer.

Comment (snoble) [reply base/src/main/scala/org/bykn/bosatsu/Macro.scala:15]: So the weird part is this works in bazel with `bazel build //cli/...` and `bazel build //core/...`... just not with `bazel build //test_workspace/...`

Comment (johnynek) [reply base/src/main/scala/org/bykn/bosatsu/Macro.scala:15]: Bazel puts external repositories on a different path. It was going to be fixed like 2 years ago and as far as I can tell they just didn’t bother. The test_workspace configured a new workspace to exercise a remote repo just for this reason.

There is a work around but I forget what it is at the moment. Ugh.

Comment (snoble) [reply base/src/main/scala/org/bykn/bosatsu/Macro.scala:15]: Ohhh... Is that what the resources_path parameters are all about?

**Comment by snoble** [comment]
I think this is ready for review. Not sure how bad the `Try` hack is

**Comment by johnynek** [comment]
I’m willing to accept the have if we can get CI green.

**Comment by snoble** [comment]
Oh, that's at least partially being held up by the version of bazel the CI is running I think

**Comment by johnynek** [comment]
merged master after bumping CI up to bazel 0.25

**Comment by johnynek** [comment]
okay, I think this is going to go green. Should we merge @snoble?

**Pull request merged**